src/wasm/wat-parser-internal.h

```diff
@@ -0,0 +1,521 @@
+
+#include <cassert>
+#include <cctype>
+#include <iostream>
+#include <optional>
+#include <variant>
+
+using namespace std::string_view_literals;
+
+namespace wasm::WATParser {
+
+namespace {
+
+// ================
+// Lexical Analysis
+// ================
+
+// The result of lexing a token fragment.
+struct LexResult {
+  std::string_view span;
+};
+
+// Lexing context that accumulates lexed input to produce a token fragment.
+struct LexCtx {
+private:
+  // The input we are lexing.
+  std::string_view input;
+
+  // How much of the input we have already lexed.
+  size_t lexedSize = 0;
+
+public:
+  explicit LexCtx(std::string_view in) : input(in) {}
+
+  // Return the fragment that has been lexed so far.
+  std::optional<LexResult> lexed() const {
+    if (lexedSize > 0) {
+      return {LexResult{input.substr(0, lexedSize)}};
+    }
+    return {};
+  }
+
+  // The next input that has not already been lexed.
+  std::string_view next() const { return input.substr(lexedSize); }
+
+  // The size of the unlexed input.
+  size_t size() const { return input.size() - lexedSize; }
+
+  // Whether there is no more input.
+  bool empty() const { return size() == 0; }
+
+  // Tokens must be separated by spaces or parentheses.
+  bool can_finish() const;
+
+  // Whether the unlexed input starts with prefix `sv`.
+  size_t starts_with(std::string_view sv) const {
+    return next().substr(0, sv.size()) == sv;
+  }
+
+  // Consume the next `n` characters.
+  void take(size_t n) { lexedSize += n; }
+
+  // Consume an additional lexed fragment.
+  void take(const LexResult& res) { lexedSize += res.span.size(); }
+
+  // Consume the prefix and return true if possible.
+  bool take_prefix(std::string_view sv) {
+    if (starts_with(sv)) {
+      take(sv.size());
+      return true;
+    }
+    return false;
+  }
+
+  // Consume the rest of the input.
+  void take_all() { lexedSize = input.size(); }
+};
+
+// The result of lexing an integer token fragment.
+struct LexIntResult : LexResult {
+  uint64_t n;
+  bool hasSign;
+};
+
+// Lexing context that accumulates lexed input to produce an integer token
+// fragment.
+struct LexIntCtx : LexCtx {
+  using LexCtx::take;
+
+private:
+  uint64_t n = 0;
+  bool hasSign = false;
+  bool negative = false;
+  bool overflow = false;
+
+  std::optional<int> getDigit(char c) {
+    if ('0' <= c && c <= '9') {
+      return {c - '0'};
+    }
+    return std::nullopt;
+  }
+
+  std::optional<int> getHexDigit(char c) {
+    if ('0' <= c && c <= '9') {
+      return {c - '0'};
+    }
+    if ('A' <= c && c <= 'F') {
+      return {10 + c - 'A'};
+    }
+    if ('a' <= c && c <= 'f') {
+      return {10 + c - 'a'};
+    }
+    return std::nullopt;
+  }
+
+public:
+  explicit LexIntCtx(std::string_view in) : LexCtx(in) {}
+
+  std::optional<LexIntResult> lexed() {
+    // Check most significant bit for underflow if we will have to negate.
+    if (overflow || (negative && (n & (1ull << 63)))) {
```

Comment:
Yes, the high bit is set in negative numbers, but here we haven't applied the sign to make the number negative yet. If the number is large enough to have its high bit set here, then once we negate it the high bit will be cleared even though the number is supposed to be negative.

Reply:
Then isn't that the same for overflow? We check `overflow` here, but it looks that variable is only set when the `uint64_t` overflows:
```cpp
        uint64_t newN = n * 10 + *d;
        if (newN < n) {
          overflow = true;
        }
```
So in case of a positive number, even if the highest bit is set, it still wouldn't overflow, but isn't that out of range?

Reply:
PTAL at the new boundary condition handling and updated tests in the latest commit.